internal/functions/serve/serve.go

```
@@ -59,179 +52,84 @@ func (i *RuntimeOption) toArgs() []string {
-		env = append(env, fmt.Sprintf("%s=%s", v.Name, v.Value))
-	}
-	return env, err
-}
+	// Create a channel to signal when a restart is needed
+	restartChan := make(chan struct{})
 
-func populatePerFunctionConfigs(cwd, importMapPath string, noVerifyJWT *bool, fsys afero.Fs) ([]string, string, error) {
-	slugs, err := deploy.GetFunctionSlugs(fsys)
-	if err != nil {
-		return nil, "", err
-	}
-	functionsConfig, err := deploy.GetFunctionConfig(slugs, importMapPath, noVerifyJWT, fsys)
-	if err != nil {
-		return nil, "", err
-	}
-	binds := []string{}
-	for slug, fc := range functionsConfig {
-		if !fc.Enabled {
-			fmt.Fprintln(os.Stderr, "Skipped serving Function:", slug)
-			continue
-		}
-		modules, err := deploy.GetBindMounts(cwd, utils.FunctionsDir, "", fc.Entrypoint, fc.ImportMap, fsys)
-		if err != nil {
-			return nil, "", err
-		}
-		binds = append(binds, modules...)
-		fc.ImportMap = utils.ToDockerPath(fc.ImportMap)
-		fc.Entrypoint = utils.ToDockerPath(fc.Entrypoint)
-		functionsConfig[slug] = fc
-		for i, val := range fc.StaticFiles {
-			fc.StaticFiles[i] = utils.ToDockerPath(val)
+	// Create a map to track watched directories
+	watchedDirs := make(map[string]bool)
+
+	// Start the file watcher in a goroutine
+	go runFileWatcher(ctx, watcher, watchedPath, restartChan, watchedDirs, fsys)
+
+	errChan := make(chan error, 1)
+
+	for {
+		select {
+		case <-ctx.Done():
+			fmt.Println("Stopping functions server...")
+			// 2. Remove existing container if any.
+			_ = utils.Docker.ContainerRemove(context.Background(), utils.EdgeRuntimeId, container.RemoveOptions{
+				RemoveVolumes: true,
+				Force:         true,
+			})
+			return ctx.Err()
+		default:
+			// Use network alias because Deno cannot resolve `_` in hostname
+			dbUrl := fmt.Sprintf("postgresql://postgres:postgres@%s:5432/postgres", utils.DbAliases[0])
+
+			serviceCancel, logsDone, err := manageFunctionServices(ctx, envFilePath, noVerifyJWT, importMapPath, dbUrl, runtimeOption, fsys, errChan)
```

Instead of returning a service cancel, can we simply call `Docker.ContainerRestart` on receiving events from the restart channel?

I believe the docker log stream will be disconnected on container restart so we can catch and ignore the error to restream.

---

ContainerRestart won't be enough to update bindings and so on, we need a full container recreation.